File: gitemail.py
from bs4 import BeautifulSoup
import requests
from typing import Optional, Union
from enum import Enum
from pprint import pprint


class Account(Enum):
    ID = "user"
    NAME = "users"


class Repo(Enum):
    ALL = None
    FORKED = True
    OWNED = False


class GitEmail():

    def __init__(self, account: str, account_type: Account = Account.NAME) -> None:
        self.api_base = "https://api.github.com"
        self.account = account
        self.account_type = account_type

    def turn_beautiful(self, response: requests.Response) -> BeautifulSoup:
        return BeautifulSoup(response, "html.parser")

    def get_response(self, url_end: str, args: dict = None) -> Optional[Union[dict, list]]:
        response = requests.get(self.api_base + url_end, params=args).json()
        try:
            response_dict = dict(response)
            if "message" in list(response_dict.keys()):
                return None
            else:
                return response_dict
        except:
            try:
                return list(response)
            except:
                return None

    def convert_to_name(self) -> Optional[str]:
        response_dict = self.get_response(f"/{self.account_type.value}/{self.account}")
        return response_dict["login"] if "login" in response_dict.keys() else None

    def get_user(self) -> Optional[dict]:
        return self.get_response(f"/{self.account_type.value}/{self.account}")

    def get_user_repos(self, repo_type: Repo = Repo.ALL, only_names: bool = False) -> Optional[list]:
        repos = self.get_response(f"/{self.account_type.value}/{self.account}/repos")
        if repo_type.value != None:
            repos = [repo for repo in repos if repo["fork"] == repo_type.value]
        if only_names:
            repos = [repo["name"] for repo in repos]
        return repos

    def get_repo_commits(self, repo: str) -> Optional[list]:
        return self.get_response(f"/repos/{self.account}/{repo}/commits")

    def git_email(self) -> str:
        if not self.get_user():
            print(f"User {self.account} could not be found.")
            return
        repos = self.get_user_repos(repo_type=Repo.OWNED, only_names=True)
        for repo in repos:
            for commit in self.get_repo_commits(repo):
                if commit["author"]["login"] == self.account or commit["author"]["id"] == self.account:
                    return commit["commit"]["author"]["email"]


if __name__ == "__main__":
    pprint(GitEmail("haraldar").git_email())

File: GitEmailExtension/lib/gitemail.js
/**
 * Contains functions that get user profile data, repositories and commits.
 */
 class GitEmail {

    constructor (account, accountType = Account.NAME) {
        this.API_BASE = "https://api.github.com";
        this.account = account;
        this.accountType = accountType;
    }
    
    /**
     * Attempts to return a jsonified response.
     * @param {string} url_end Basically the url suffix that contains the parameters.
     * @returns The returned json from the GitHub API.
     */
    async getResponse (url_end) {
        const response = await fetch(
            this.API_BASE + "/" + url_end,
            {
                method: 'GET',
                headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                }
            });
            return response.json();
    }

    /**
     * Sets the object's account and account type.
     * @param {String} account The account name or id to set.
     * @param {Repo} accountType Defines the type of account.
     * @returns The account name.
     */
    setAccount (account, accountType) {
        this.account = account;
        this.accountType = accountType;
        return account;
    }

    /**
     * Converts the objects' account from Account.ID to Account.Name.
     * @returns The username or null.
     */
    async convertToName () {
        const response = await this.getUser();
        return (user != null) ?
            this.setAccount(response["login"], Account.NAME) :
            null;
    }

    /**
     * Fetches the user account JSON from the GitHub API.
     * @returns The user JSON.
     */
    async getUser () {
        const user = await this.getResponse(`${this.accountType}/${this.account}`);
        return (user.hasOwnProperty("message")) ? null : user;
    }

    /**
     * Fetches the users' repositories (code piece created with the help of @nikitakhutorni).
     * @param {Repo} repoType The type filter for accepted repositories.
     * @param {Boolean} onlyNames Converts the results strictly to their repository name.
     * @returns A list of repositories. Either each entry is JSON-type or a string.
     */
    async getUserRepos (repoType = Repo.ALL, onlyNames = false) {
        const response = await this.getResponse(`${this.accountType}/${this.account}/repos`);
        return (response.hasOwnProperty("message")) ?
            null :
            response.filter(repo => repoType == Repo.ALL || repoType == repo["fork"])
            .map(repo => onlyNames ? repo.name : repo);
    }

    /**
     * Fetches a specific repositories' commits for the objects' user.
     * @param {String} repo The repository to get the commits for.
     * @returns A list of commits. Maximum amount is 30 due to public API limitations.
     */
    async getRepoCommits (repo) {
        return await this.getResponse(`repos/${this.account}/${repo}/commits`);
    }

    async getRepoLanguages () {

    }

    /**
     * Determines the users public committing email.
     * @returns The users' public email, true if no repos to scan, false if no repos
     * that contain the user's commits or null if user doesn't exist.
     */
    async gitEmail () {
        const user = await this.getUser();
        if (user == null) return null;
        const repos = await this.getUserRepos(Repo.OWNED, true);
        if (repos.length == 0) {
            return true;
        }
        else {
            for (let repo in repos) {
                let commits = await this.getRepoCommits(repos[repo]);
                for (var commit in commits) {
                    if (commits[commit].author != null) {
                        if (commits[commit].author.login == this.account) {
                            return commits[commit].commit.author.email;
                        }
                    }
                }
            }
            return false;
        }
    }
}


/**
 * A restricted amount of account types.
 */
class Account {

    static ID = "user";
    static NAME = "users";

    constructor (accountType) {
        this.accountType = accountType;
    }

    toString () {
        return this.accountType;
    }

}

/**
 * A restricted amount of repository types.
 */
class Repo {

    static ALL = null;
    static FORKED = true;
    static OWNED = false;

    constructor (repo_type) {
        this.repo_type = repo_type;
    }

    toBoolean () {
        return this.repo_type;
    }
}

export {GitEmail, Account, Repo};